Url validation and Authorisation

Nimmyjv opened the pull request.

NimmyVipin committed:
Add validators folder, url validation

NimmyVipin committed:
Authorize collection added by each user with pundit gem

redtachyons commented on app/policies/collection_policy.rb:
```diff
@@ -0,0 +1,6 @@
+     def destroy?
+      record.user == user
+     end
+end
```
Indentation

redtachyons commented on config/locales/en.yml:
```diff
@@ -21,3 +21,4 @@
 
 en:
   hello: "Hello world"
+  
```
why?

nithinbekal replied:
Rubocop warns about not having newline at end of file. Either way, the last line shouldn't have the indentation.

nithinbekal commented on app/services/collections/delete.rb:
```diff
@@ -19,6 +19,7 @@ def call
     attr_reader :user, :collection
 
     def delete_collection
+      Pundit.authorize(user, collection, :destroy?)
```
This should be done in the controller, not here.

nithinbekal commented on app/services/websites/create.rb:
```diff
@@ -1,3 +1,4 @@
+# frozen_string_literal: true
```
Not needed. Ignore this warning if rubocop shows it.

nithinbekal commented on app/validators/url_validator.rb:
```diff
@@ -0,0 +1,24 @@
+    false
+  end
+
+  def valid_host? host
```
Parentheses around args.

redtachyons replied:
Afaik rubocop suggests to not use  parenthesis if the number of arguments are 1

nithinbekal replied:
I doubt that. Rubocop follows [bbatsov's styleguide](https://github.com/bbatsov/ruby-style-guide), which prefers parentheses all the time, and I don't remember seeing a warning about this even though I always use parens around single args.

Maybe you're thinking about the case where there are no args?

nithinbekal commented on app/validators/url_validator.rb:
```diff
@@ -0,0 +1,24 @@
+    host.present? && valid_characters?(host)
+  end
+
+  def valid_characters? host
```
Parentheses around args.

nithinbekal commented on app/models/website.rb:
```diff
@@ -1,9 +1,10 @@
+# frozen_string_literal: true
 class Website < ApplicationRecord
+  before_validation :add_protocol_to_website 
+  validates :url, url: true
```
Group different types of statements (callbacks, validations, associations, etc) together, and separate them by empty lines.

nithinbekal commented:
@Nimmyjv When you submit a PR that fixes a specific issue, please add "fixes #issueNo" in the description. It will be automatically connected to it.

NimmyVipin committed:
Modify as run pronto

NimmyVipin committed:
Authorization moved from service to controller

nithinbekal commented on app/models/website.rb:
```diff
@@ -38,6 +40,6 @@ def fetch_meta_description
 
   def add_protocol_to_website
-    self.url = (self.url[/^http:\/\//] || self.url[/^https:\/\//]) ? self.url : "http://#{self.url}"
+    self.url = url[/^http:\/\//] || url[/^https:\/\//] ? url : "http://#{url}"
```
This is difficult to read. In Leo, we rewrote it like this:

```ruby
    def add_protocol_to_url
      return if protocol_present?
      self.url = "https://#{url}"
    end

    def protocol_present?
      return false if url.blank?
      url[/^http:\/\//] || url[/^https:\/\//]
    end
```

nithinbekal commented on app/validators/url_validator.rb:
```diff
@@ -1,3 +1,4 @@
+
```
Blank line not needed.

nithinbekal commented on app/policies/collection_policy.rb:
```diff
@@ -1,6 +1,6 @@
+# frozen_string_literal: true
```
As mentioned before, remove this and ignore frozen_string_literal warnings from rubocop. Also, edit rubocop.yml to ignore these warnings.

Nimmyjv replied:
ok

nithinbekal commented on app/controllers/application_controller.rb:
```diff
@@ -1,3 +1,4 @@
+# frozen_string_literal: true
```
Remove this.

NimmyVipin committed:
Edit rubocop.yml to avoid frozenstringliteralwarnings

NimmyVipin committed:
Modify method to add protocol to url

NimmyVipin committed:
Fix rubocop warnings

NimmyVipin committed:
Correct method name

nithinbekal commented on app/controllers/collections_controller.rb:
```diff
@@ -7,7 +7,9 @@ def create
   def destroy
-    Collections::Delete.call(params[:id], current_user)
+    collection = Collection.find(params[:id])
+    Pundit.authorize(current_user, collection, :destroy?)
```
Is `Pundit.` required? `authorize(...)` works, I believe.

nithinbekal commented on app/services/collections/delete.rb:
```diff
@@ -19,7 +19,7 @@ def call
 
     def delete_collection
-      CollectionWebsite.where('collection_id = ?', collection.id).destroy_all
+      CollectionWebsite.where(collection_id: collection.id).destroy_all
```
Can't this be achieved by doing `dependent: destroy` on `has_many :collection_websites`?

NimmyVipin committed:
Remove query to delete from join table

NimmyVipin committed:
Modify authorize method

NimmyVipin committed:
change https to http

nithinbekal commented:
LGTM. 👍

redtachyons commented on app/controllers/application_controller.rb:
```diff
@@ -2,10 +2,16 @@ class ApplicationController < ActionController::Base
+
+  def user_not_authorized
+    flash[:alert] = 'You are not authorized to perform this action.'
+    redirect_to(request.referer || root_path)
```
You can write this like     ```redirect_back(fallback_location: root_path)```

Nimmyjv replied:
ok

NimmyVipin committed:
Destroy collectionWebsites also when collection deleted

NimmyVipin committed:
To check if same url is added to a single list.For eg; earlier when redpanthers.co is added it is stored as http://redpanthers.co
When again redpanthers.co added to the same list, it was added because it is compared with http://redpanthers.co
To prevent this, when url is added itself, it is checked whether starts with http or https. Else, http is added and then compared

NimmyVipin committed:
In Rails 5, redirect_to :back has been deprecated and instead a new method has been added called redirect_back.

NimmyVipin committed:
Add tool tips in multiple list page

nithinbekal commented on app/assets/javascripts/custom.js:
```diff
@@ -1,4 +1,4 @@
-$(document).ready(function() {
+$(document).ready(function() { 
```
Trailing space.

nithinbekal commented on app/models/invite.rb:
```diff
@@ -1,5 +1,4 @@
 class Invite < ApplicationRecord
   has_secure_token
-
```
Leave a blank line between different kind of declarations.

nithinbekal commented:
@Nimmyjv Are the tests running fine? I think these changes might be breaking some of the tests, or causing HTTP requests. If there are HTTP requests to 3rd parties, use [webmock](https://github.com/bblimke/webmock) to stub out the requests.

NimmyVipin committed:
Edit rubocop file to ignore regex literal

NimmyVipin committed:
change redirection path

NimmyVipin committed:
Spacing correction

NimmyVipin committed:
Solve error while run rake db:seed

NimmyVipin committed:
Modify test file to fix failures

nithinbekal commented on app/models/invite.rb:
```diff
@@ -1,5 +1,5 @@
 class Invite < ApplicationRecord
   has_secure_token
-
+  
```
Trailing whitespace.

nithinbekal commented on app/views/static_pages/_collection.html.erb:
```diff
@@ -96,4 +91,4 @@
 
-        
\ No newline at end of file
+       
```
Trailing whitespace, and unnecessary 2nd blank line.

nithinbekal commented on app/services/websites/create.rb:
```diff
@@ -14,9 +14,15 @@ def initialize(params, user)
+              url
+            else
+              "http://#{url}"
+            end
```
Extract the condition out, so you can rewrite it like this:

```ruby
url = "http://#{url}" unless protocol_given?(url)
```

nithinbekal commented on db/seeds.rb:
```diff
@@ -20,4 +25,4 @@
 end
 
-AdminUser.create!(email: 'admin@example.com', password: 'password', password_confirmation: 'password')
+AdminUser.first_or_create!(email: 'admin@example.com', password: 'password', password_confirmation: 'password')
```
Do we really need `first_or_create` in seeds.rb? It is only run in an empty database, right?

nithinbekal commented on db/seeds.rb:
```diff
@@ -5,11 +5,16 @@
 
-urls.each do |url|
-  user.websites.create!(url: url)
+user = User.first_or_create!(name: 'abc', email: 'abc@mail.com', password: '123456',                    password_confirmation: '123456')
```
Whitespace in the middle of line.

nithinbekal commented on db/seeds.rb:
```diff
@@ -5,11 +5,16 @@
+  urls.each do |url|
+    collection.websites.create!(url: url)
+  end
 end
```
In seeds.rb, you don't really need all this complexity, since you know what data you want to add. You can write it like this:

```ruby
user = User.create!(name: 'abc', email: 'abc@mail.com', password: '123456', password_confirmation: '123456')
collection = user.collections.create!(name: 'First Collection')
collection.websites.create!(url: 'https://redpanthers.co')
```

When adding seed data, it would be good to add valid url, since you will be using it in development.

nithinbekal commented on test/services/websites/create_test.rb:
```diff
@@ -6,32 +6,47 @@ class CreateTest < ActiveSupport::TestCase
+      Websites::Create.call(params, user)
+
+      user.reload
+      assert_equal 1, user.site_number, 'site number'
```
Does this field contain number of websites a user has? Then can we call it `websites_count` instead? This seems an awkward name for the field.

(Do this in a separate PR rather than mixing these.)

NimmyVipin committed:
Run pronto

NimmyVipin committed:
Code refactored

NimmyVipin committed:
Indentation corrected

NimmyVipin committed:
change name

nithinbekal commented on app/services/websites/create.rb:
```diff
@@ -54,5 +50,9 @@ def fetch_rank
+
+    def protocol_given?(url)
+      url if (url.start_with? 'http://') || (url.start_with? 'https://')
+    end
```
Don't return `url`. A method ending with `?` should return true or false.

nithinbekal commented on app/views/static_pages/_collection.html.erb:
```diff
@@ -31,64 +31,56 @@
+            <% end %>
+          <% end %>
+          <% if collection_nil.empty? %>
+            <% @collection_names.any? %>
```
What does this do? It just seems to return true or false and render nothing.

redtachyons replied:
:joy:

nithinbekal commented on test/services/websites/create_test.rb:
```diff
@@ -11,7 +11,7 @@ class CreateTest < ActiveSupport::TestCase
 
       user.reload
-      assert_equal 1, user.site_number, 'site number'
+      assert_equal 1, user.websites.count, 'websites_count'
```
I meant renaming `site_number` to `websites_count` since that is more accurately describing what it contains, not `user.websites.count`.

Nimmyjv replied:
okay

NimmyVipin committed:
remove return value
